R/utilities.R

```diff
@@ -127,19 +127,113 @@ get_ssh_session = function(host="gphost01.bcgsc.ca"){
+  }else if(return_as == "df"){
+    region = dplyr::select(gene_coordinates, chromosome, start, end, gene_name, hugo_symbol, ensembl_gene_id) %>%
+      as.data.frame() %>%
+      dplyr::arrange(chromosome, start)
   }
+
   if(!missing(gene_symbol)){
-    gene_coordinates = dplyr::filter(grch37_gene_coordinates, hugo_symbol == gene_symbol)
+    message(paste0(nrow(region), " region(s) returned for ", length(gene_symbol), " gene(s)"))
+  }
+
+  if(!missing(ensembl_id)){
+    message(paste0(nrow(region), " region(s) returned for ", length(ensembl_id), " gene(s)"))
   }
-  region = paste0(gene_coordinates$chromosome, ":", gene_coordinates$start, "-", gene_coordinates$end)
+
   return(region)
 }
 
+#' Return gennes residing in defined region(s)
+#'
+#' @param region Regions to intersect genes with, this should be a bed-like df (chromosome, start, end).
+#' @param gene_format Parameter for specifying the format of returned genes, default is "hugo", other acceptable inputs are "ensembl".
+#' @param genome_build Reference genome build.
+#' @param chr_select Optional parameter to subset plot to specific chromosomes. Default value is chr1-22.
+#'
+#' @import data.table
+#' @return
+#' @export
+#'
+#' @examples
+#' myc_region = gene_to_region(gene_symbol = "MYC", genome_build = "grch37", return_as = "df")
+#' region = region_to_gene(region = myc_region, gene_format = "hugo", genome_build = "grch37")
+#'
+region_to_gene = function(region,
+                          gene_format = "hugo",
+                          genome_build = "grch37",
+                          chr_select = paste0("chr", c(1:22))){
+
+  if(genome_build == "grch37"){
+    gene_list = grch37_gene_coordinates
+  }
+
+  if(genome_build == "hg38"){
+    gene_list = hg38_gene_coordinates
+  }
+
+  #transform regions to data tables
+  region_table = as.data.table(region)
+  gene_table = as.data.table(gene_list)
+
+  #set keys
+  data.table::setkey(region_table, chromosome, start, end)
+  data.table::setkey(gene_table, chromosome, start, end)
+
+  #intersect regions
+  intersect = data.table::foverlaps(region_table, gene_table, nomatch = 0)
+  colnames(intersect)[7] = "region_start"
+  colnames(intersect)[8] = "region_end"
+
+  #transform object to data frame
+  inter_df = as.data.frame(intersect)
+
+  #organize columns to match the expected format
+  if(gene_format == "hugo"){
+    genes = select(inter_df, chromosome, start, end, hugo_symbol, region_start, region_end)
+  }else if(gene_format == "ensembl"){
+    genes = select(inter_df, chromosome, start, end, ensembl_gene_id, region_start, region_end)
+  }
+
+  #paste chr in chromosome column, if not there
+  if(!str_detect(genes$chromosome, "chr")){
+    genes = mutate(genes, chromosome = paste0("chr", chromosome))}
+
+  genes = genes[genes$chromosome %in% chr_select, ]
+
+  genes = as.data.frame(genes) %>%
+      dplyr::arrange(chromosome, start)
+
+  message(paste0(nrow(genes), " gene(s) returned for ", nrow(region), " region(s)"))
+
```

Comment:
Does this return the outer coordinates of each gene in all the regions provided? I can't tell from the documentation

Reply:
Yes, together with the regions provided as such;
```
> region_to_gene(region = myc_region, gene_format = "hugo", genome_build = "grch37")
1 gene(s) returned for 1 region(s)
  chromosome     start       end hugo_symbol region_start region_end
1       chr8 128747680 128753674         MYC    128747680  128753674
```

Reply:
Is it a coincidence that region_start and start are identical and region_end and end are identical? What happens if you specify a larger region as input? What was myc_region here?

Reply:
For this example I first ran gene_to_region and then gave that output to region_to_gene, as such;
```
> myc_region = gene_to_region(gene_symbol = "MYC", genome_build = "grch37", return_as = "df")
1 region(s) returned for 1 gene(s)
> region_to_gene(region = myc_region, gene_format = "hugo", genome_build = "grch37")
1 gene(s) returned for 1 region(s)
  chromosome     start       end hugo_symbol region_start region_end
1       chr8 128747680 128753674         MYC    128747680  128753674
```

Here's an example with a larger region (p-arm chromosome 1);
```
> some_region = data.frame(chromosome = "1", start = 10000, end = 121500000)
> region_to_gene(region = some_region, gene_format = "ensembl", genome_build = "grch37")
2785 gene(s) returned for 1 region(s)
    chromosome   start     end ensembl_gene_id region_start region_end
1         chr1   11869   14412 ENSG00000223972        10000  121500000
2         chr1   14363   29806 ENSG00000227232        10000  121500000
3         chr1   29554   31109 ENSG00000243485        10000  121500000
4         chr1   29554   31109 ENSG00000243485        10000  121500000
5         chr1   29554   31109 ENSG00000243485        10000  121500000
```
(only printed the first 5 rows, there are 2785 genes returned for the specified region)

Reply:
Why are lines 3, 4 and 5 identical?

Reply:
According to what's in the `grch37_gene_coordinates` R data file, these MIR1302 genes have the same ensemble ID;
```
> grch37_sub = dplyr::filter(grch37_gene_coordinates, chromosome == "1") %>% 
+   arrange(start) %>%
+   head(5) %>%
+   print()
# A tibble: 5 × 6
  ensembl_gene_id chromosome start   end gene_name  hugo_symbol
  <chr>           <chr>      <dbl> <dbl> <chr>      <chr>      
1 ENSG00000223972 1          11869 14412 DDX11L1    DDX11L1    
2 ENSG00000227232 1          14363 29806 WASH7P     WASH7P     
3 ENSG00000243485 1          29554 31109 MIR1302-10 MIR1302-11 
4 ENSG00000243485 1          29554 31109 MIR1302-10 MIR1302-10 
5 ENSG00000243485 1          29554 31109 MIR1302-10 MIR1302-9  
```

the same snippet but with hugo symbols looks like this;
```
> region_to_gene(region = some_region, gene_format = "hugo", genome_build = "grch37")
2785 gene(s) returned for 1 region(s)
    chromosome   start     end hugo_symbol region_start region_end
1         chr1   11869   14412     DDX11L1        10000  121500000
2         chr1   14363   29806      WASH7P        10000  121500000
3         chr1   29554   31109  MIR1302-11        10000  121500000
4         chr1   29554   31109  MIR1302-10        10000  121500000
5         chr1   29554   31109   MIR1302-9        10000  121500000
```

Reply:
Got it. I think that since what we return doesn't always contain the same columns the data frame we return should be collapsed to unique rows. Does that make sense?

Reply:
This has been implemented.